src/main/java/org/broadinstitute/hellbender/tools/exome/IntervalBackedExonDB.java

```diff
@@ -0,0 +1,361 @@
+            return searchIndex;
+    }
+
+    /**
+     * Special case for {@link #index(GenomeLoc)} for {@link GenomeLoc#WHOLE_GENOME}
+     * @param location the whole-genome location.
+     * @return never {@code null}.
+     */
+    private int indexWholeGenome(final GenomeLoc location) {
+        final int size = sortedIntervals.size();
+        if (size == 0)
+            return -1;
+        else if (size == 1)
+            return 0;
+        else
+            throw new AmbiguousExonException("" + location + " overlaps with at least 2: "
+                    + sortedIntervals.get(0) + " and " + sortedIntervals.get(1));
+    }
+
+    @Override
+    public List<GenomeLoc> exons(final GenomeLoc overlapRegion) {
+        if (overlapRegion == GenomeLoc.WHOLE_GENOME)
+            return Collections.unmodifiableList(sortedIntervals);
+        final int searchIndex = cachedBinarySearch(overlapRegion);
+        if (searchIndex < 0)
+            return Collections.emptyList();
+        final int firstMatchingIndex = extendSearchIndexBackwards(overlapRegion, searchIndex);
+        final int lastMatchingIndex = extendSearchIndexForward(overlapRegion, searchIndex);
+        if (firstMatchingIndex == lastMatchingIndex)
+            return Collections.singletonList(sortedIntervals.get(searchIndex));
+        else
+            return Collections.unmodifiableList(sortedIntervals.subList(firstMatchingIndex, lastMatchingIndex + 1));
+    }
+
+    private int lastBinarySearchResult = -1;
+
+    private int cachedBinarySearch(final GenomeLoc location) {
+
+        if (lastBinarySearchResult < 0) {
+            final int candidate = - lastBinarySearchResult - 1;
+            if (candidate >= sortedIntervals.size())
+                return lastBinarySearchResult = Collections.binarySearch(sortedIntervals,location,OVERLAP_SEARCH_COMPARATOR);
+            else if (sortedIntervals.get(candidate).overlapsP(location))
+                return lastBinarySearchResult = candidate;
+            else
+                return lastBinarySearchResult = Collections.binarySearch(sortedIntervals,location,OVERLAP_SEARCH_COMPARATOR);
+        } else {
+            if (sortedIntervals.get(lastBinarySearchResult).overlapsP(location))
+                return lastBinarySearchResult;
+            else {
+                final int candidate = lastBinarySearchResult + 1;
+                if (candidate == sortedIntervals.size())
+                    return lastBinarySearchResult = Collections.binarySearch(sortedIntervals,location,OVERLAP_SEARCH_COMPARATOR);
+                if (sortedIntervals.get(candidate).overlapsP(location))
+                    return lastBinarySearchResult = candidate;
+                else
+                    return lastBinarySearchResult = Collections.binarySearch(sortedIntervals,location,OVERLAP_SEARCH_COMPARATOR);
+            }
+        }
+    }
+
+    @Override
+    public List<GenomeLoc> exons(final int from, final int to) {
+        if (from < 0)
```

Comment:
premature optimization. Keep the human in mind, not the machine. 1 case is much easier to read than 3